### PR TITLE
Create `water` schema views

### PR DESCRIPTION
**Pull request by Jozzey**
https://eaflood.atlassian.net/browse/WATER-4188

As part of the work we have been doing on two-part tariff we are going to be creating all our new tables in a single schema `public`.

We have also decided that when there is a legacy table that we are still going to need for SROC. We are going to create a View of that table in the `public` schema and correct any issues with naming conventions, unused fields etc in the view. The first of these views that represents the old `returns` tables have been created in https://github.com/DEFRA/water-abstraction-system/pull/531

In this PR we will build the migration scripts to create the Views we need for SROC in the `water` schema.

### CODE REVIEW

File: db/migrations/public/20231127212015_create-charge-references-view.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        'is_section_127_agreement_enabled',
+        'scheme',
+        'is_restricted_source',
+        'water_model',
+        'volume',
+        'billing_charge_category_id AS charge_category_id',
+        'additional_charges',
+        'adjustments',
+        'is_section_126_agreement_enabled',
+        'is_section_130_agreement_enabled',
```
Comment (Cruikshanks) [inline db/migrations/public/20231127212015_create-charge-references-view.js:41]: Same as previous comment

```suggestion
        'is_section_127_agreement_enabled AS section_127_agreement',
        'scheme',
        'is_restricted_source AS restricted_source',
        'water_model',
        'volume',
        'billing_charge_category_id AS charge_category_id',
        'additional_charges',
        'adjustments',
        'is_section_126_agreement_enabled AS section_126_agreement',
        'is_section_130_agreement_enabled AS section_130_agreement',

```

File: db/migrations/public/20231127214320_create-charge-versions-view.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        'licence_ref',
+        'scheme',
+        'external_id',
+        'version_number',
+        'start_date',
+        'status',
+        'apportionment',
+        'error',
+        'end_date',
+        'billed_upto_date',
```
Comment (Cruikshanks) [inline db/migrations/public/20231127214320_create-charge-versions-view.js:21]: Upto isn't a word so we can correct this in the view.

```suggestion
        'billed_upto_date AS billed_up_to_date',

```

File: db/migrations/public/20231127222209_create-licences-view.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+exports.up = function (knex) {
+  return knex
+    .schema
+    .createView(viewName, (view) => {
+      // NOTE: We have commented out unused columns from the source table
+      view.as(knex('licences').withSchema('water').select([
+        'licence_id AS id',
+        'region_id',
+        'licence_ref',
+        'is_water_undertaker',
```
Comment (Cruikshanks) [inline db/migrations/public/20231127222209_create-licences-view.js:14]: Same as previous comment

```suggestion
        'is_water_undertaker AS water_undertaker',

```

File: db/migrations/public/20231127222209_create-licences-view.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        'regions',
+        'start_date',
+        'expired_date',
+        'lapsed_date',
+        'revoked_date',
+        'suspend_from_billing',
+        // 'is_test',
+        'include_in_supplementary_billing',
+        'include_in_sroc_supplementary_billing',
```
Comment (Cruikshanks) [inline db/migrations/public/20231127222209_create-licences-view.js:23]: Not sure this suggestion is necessarily better. But can't afford to miss the opportunity to make these more usable in our code! How about

```suggestion
        'include_in_supplementary_billing AS include_in_presroc_billing',
        'include_in_sroc_supplementary_billing AS include_in_sroc_billing',

```

File: db/migrations/public/20231127222742_create-purposes-view.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  return knex
+    .schema
+    .createView(viewName, (view) => {
+      // NOTE: We have commented out unused columns from the source table
+      view.as(knex('purposes_uses').withSchema('water').select([
+        'purpose_use_id AS id',
+        'legacy_id',
+        'description',
+        'loss_factor',
+        'is_two_part_tariff',
```
Comment (Cruikshanks) [inline db/migrations/public/20231127222742_create-purposes-view.js:15]: Same as previous comment

```suggestion
        'is_two_part_tariff AS two_part_tariff',

```

File: db/migrations/public/20231127224323_create-transactions-view.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        'scheme',
+        'aggregate_factor',
+        'adjustment_factor',
+        'charge_category_code',
+        'charge_category_description',
+        'is_supported_source',
+        'supported_source_name',
+        'is_water_company_charge',
+        'is_winter_only',
+        'is_water_undertaker',
```
Comment (Cruikshanks) [inline db/migrations/public/20231127224323_create-transactions-view.js:58]: Same comment as previous

```suggestion
        'is_credit AS credit',
        'charge_type',
        'authorised_quantity',
        'billable_quantity',
        'authorised_days',
        'billable_days',
        'status',
        'description',
        'external_id',
        'volume',
        'section_126_factor',
        'section_127_agreement',
        'section_130_agreement',
        'is_de_minimis AS deminimis',
        'is_new_licence AS new_licence',
        'legacy_id',
        'metadata',
        'source_transaction_id',
        'calc_source_factor',
        'calc_season_factor',
        'calc_loss_factor',
        'calc_suc_factor',
        'calc_s_126_factor',
        'calc_s_127_factor',
        'calc_eiuc_factor',
        'calc_eiuc_source_factor',
        'is_credited_back AS credited_back',
        'is_two_part_second_part_charge AS second_part_charge',
        'scheme',
        'aggregate_factor',
        'adjustment_factor',
        'charge_category_code',
        'charge_category_description',
        'is_supported_source AS supported_source',
        'supported_source_name',
        'is_water_company_charge AS water_company_charge',
        'is_winter_only AS winter_only',
        'is_water_undertaker AS water_undertaker',

```

File: db/migrations/public/20231127230416_create-workflows-view.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    .createView(viewName, (view) => {
+      view.as(knex('charge_version_workflows').withSchema('water').select([
+        'charge_version_workflow_id AS id',
+        'licence_id',
+        'created_by',
+        'approver_comments',
+        'status',
+        'data',
+        'licence_version_id',
+        'date_deleted',
```
Comment (Cruikshanks) [inline db/migrations/public/20231127230416_create-workflows-view.js:17]: What do you think about making this consistent with the other fields?

```suggestion
        'date_deleted AS deleted_at',

```

**Review by Cruikshanks — CHANGES_REQUESTED** [review]
Think there are 1 couple of corrections in here, some suggestions for alternate names.

The key change is dropping the convention of prefixing boolean fields with `is_`, especially as going through this, I have seen they haven't followed the convention of applying it consistently.

File: db/migrations/public/20231127193216_create-bills-view.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        'billing_batch_id AS bill_run_id',
+        'financial_year_ending',
+        'invoice_number',
+        'legacy_id',
+        'metadata',
+        'credit_note_value',
+        'invoice_value',
+        'is_de_minimis',
+        'external_id',
+        'is_flagged_for_rebilling',
```
[the remark this answers is not in the document]
Comment (Jozzey) [reply db/migrations/public/20231127193216_create-bills-view.js:25]: 💯

**Review by Cruikshanks — APPROVED** [review]

**Pull request merged**